library/std/src/alloc.rs

```diff
@@ -127,6 +127,122 @@ pub use alloc_crate::alloc::*;
+// implementation.
+#[cfg(not(debug_assertions))]
+#[stable(feature = "alloc_system_type", since = "1.28.0")]
+unsafe impl GlobalAlloc for System {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        unsafe { Imp.alloc(layout) }
+    }
+
+    #[inline]
+    unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+        unsafe { Imp.alloc_zeroed(layout) }
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        unsafe { Imp.dealloc(ptr, layout) }
+    }
+
+    #[inline]
+    unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
+        unsafe { Imp.realloc(ptr, layout, new_size) }
+    }
+}
+
+// Some system allocators (most notably any provided by calling malloc) will always return pointers
+// with an alignment of 8. So for any allocation with an alignment less than 8, we increase the
+// alignment to 8 and return a pointer which is offset into the allocation such that it is not
+// over-aligned.
+// We always bump up the size of an allocation by 8 when the alignment is less than 8.
+#[cfg(debug_assertions)]
+trait LayoutExt {
+    fn with_alignment_padding(self) -> Self;
+    unsafe fn add_alignment_padding(self, ptr: *mut u8) -> *mut u8;
+    unsafe fn remove_alignment_padding(self, ptr: *mut u8) -> *mut u8;
+}
+#[cfg(debug_assertions)]
+impl LayoutExt for Layout {
+    fn with_alignment_padding(self) -> Self {
+        if self.align() < 8 {
+            Layout::from_size_align(self.size() + (8 - self.align()), 8).unwrap()
+        } else {
+            self
+        }
+    }
+
+    unsafe fn add_alignment_padding(self, ptr: *mut u8) -> *mut u8 {
+        if !ptr.is_null() && self.align() < 8 {
+            // SAFETY: This must be called on a pointer previously returned by a padded Layout,
+            // which will always have space to do this offset
+            unsafe { ptr.add(8 - self.align()) }
+        } else {
+            ptr
+        }
+    }
+
+    unsafe fn remove_alignment_padding(self, ptr: *mut u8) -> *mut u8 {
+        // We cannot just do the inverse of add_alignment_padding, because if a user deallocates
+        // with the wrong Layout, we would use that wrong Layout here to deduce the wrong offset to
+        // remove from the pointer. That would turn code that works fine because the underlying
+        // allocator ignores the Layout (but is technically UB) into code which does worse UB or
+        // halts the program with an unhelpful diagnostic from the underlying allocator.
+        // So we have two reasonable options. We could detect and clearly report the error
+        // ourselves, or since we know that all our alignment adjustments involve the low 3 bits,
+        // we could clear those and make this allocator transparent.
+        // At the moment we do the latter because it is unclear how to emit an error message from
```

Comment:
IMO this really needs to report the issue/abort/something. What it does now seems to just be returning a misaligned pointer silently, and hoping the caller notices. This seems likely to end up not noticed at all, or noticed far away from the allocation site.

I don't have strong thoughts on how to handle it, but I think printing output to stderr and aborting might be a start (or perhaps seeing what `handle_alloc_error` does? Does it have some solution for the "report-and-abort without unwinding" problem?).

Given that this is mostly the scaffolding for other checks... I think it's worth figuring this out sooner rather than later.

Reply:
Also to be clear, even in `std`, you are correct that there may be no standard error stream (such as on `wasm*-unknown-unknown`). I think it's okay for that to be best effort, so long as the effort is actually pretty good and usually does help the user, (although the aborting should be unconditional, IMO).

And it's a debugging feature, one that is only available via unstable flags. It not being possible on some targets doesn't mean it can't be available on any. We can always just not enable this on targets where we'd have no method of doing anything about detected misuse. That said, I think just about everything with `std` should be able to do something here (perhaps wasm has no stderr, though).

Reply:
> What it does now seems to just be returning a misaligned pointer silently, and hoping the caller notices.

I'm pretty sure this doesn't return a misaligned pointer silently. The effect of this strategy is to populate the lowest 3 bits to the greatest extent possible on the application side, and on the actual system allocator side keep them always clear. So this just indiscriminately clears the lowest 3 bits when handing a pointer back to the system allocator, which I'm pretty sure makes deallocation with the wrong layout totally fine if the underlying system allocator also doesn't care about layout.